### ISSUE TITLE
Cop idea: Not enough arguments for enum_for/to_enum

### ISSUE THREAD

**Issue by pocke** [rubocop-hq/rubocop#7753]
## Is your feature request related to a problem? Please describe.

We have an idiom of `enum_for`/`to_enum` method for defining an iterator.

```ruby
class C
  def each_item(&block)
    return enum_for(__method__) unless block_given?

    iterate_someting_with_block block
  end
end

c = C.new

# We can call C#each with and without a block.
c.each_item do |item|
end
c.each_item.map do |item|
end
```

It is a useful idiom! But we sometimes make a bug if the iterator receives arguments. For example:

```ruby
class C
  def each_item(type, &block)
    return enum_for(__method__) unless block_given?

    iterate_something_with_block type, block
  end
end
c = C.new

# It works well.
c.each_item(:foo) do |item|
end

# But it does not work well because enum_for should receive `type` but actually doesn't.
c.each_item(:foo).map do |item|
end
```

It should be `return enum_for(__method__, type) unless block_given?`.

In this case, it is not a serious problem because `c.each_item(:foo).map{}` raises an `ArgumentError`.
But it will be serious when the argument is optional, like `def each_item(type = :default, &block)`.
It ignores `type` argument silently.

## Describe the solution you'd like

Add a new cop to detect this problem.

## Describe alternatives you've considered

Nothing

## Additional context

* to_enum is an alias of enum_for
* Probably we should focus the idiom, it means the cop should accept any `enum_for` without `return enum_for(__method__) unless block_given?` form.

**Comment by marcandre** [comment]
I've made that mistake myself a few times.

Cop should only apply when: the receiver is `nil` or `self`.
Maybe it should check that either the method name (in your example `:each_item`) or `__method__` is used as the first argument too.
